src/samVisualize/rendering/forceDirectedGraph.ts

```diff
@@ -0,0 +1,703 @@
+     * Appends an `svg` element to the `body` of a given `Document`
+     
+     * @param width The width of the `svg` element
+     * @param height The height of the `svg` element
+     * @param id The id of the svg created
+     * @param doc The document whose `body` the `svg` element is added to
+     * @returns A d3 `Selection` of the `svg` element
+     */
+    public constructSVG(
+        width: number,
+        height: number,
+        id?: string, // Used to select and test created elements. No purpose outside test environment.
+        doc?: Document
+    ): d3.Selection<SVGSVGElement, unknown, null | HTMLElement, unknown> {
+        // This document param is necessary only for testing, to give a 'fake' DOM for d3 to manipulate
+        const body: d3.Selection<SVGSVGElement, unknown, null | HTMLElement, unknown> = doc
+            ? d3.select(doc).select('body')
+            : d3.select('body')
+        const svg = body.append<SVGSVGElement>('svg').attr('width', width).attr('height', height)
+        if (id) {
+            svg.attr('id', id)
+        }
+        return svg
+    }
+
+    /**
+     * Defines an arrowhead `marker` element, with a given color.
+     * @param svg A d3 `Selection` of an `svg` element to which the `marker` is added
+     * @param color The fill color of the `marker` element
+     * @param id The id attribute of the `marker` element
+     */
+    public defineArrowHead(
+        svg: d3.Selection<SVGSVGElement, unknown, null | HTMLElement, unknown>,
+        id?: string // Used to select and test created elements. No purpose outside test environment.
+    ): void {
+        // Create marker within defs
+        svg.append('defs')
+            .append('marker')
+            .attr('class', 'arrowhead')
+            .attr('id', id ? id : this.arrowheadID) // Use test id if present
+            .attr('viewBox', RenderConstants.arrowheadViewbox)
+            // refX = 0 to draw the arrowhead at the end of the path
+            .attr('refX', 0)
+            // refY = 0 to draw the arrowhead centered on the path
+            .attr('refY', 0)
+            .attr('orient', 'auto')
+            .attr('markerWidth', RenderConstants.arrowheadSize)
+            .attr('markerHeight', RenderConstants.arrowheadSize)
+            .attr('markerUnits', 'userSpaceOnUse')
+            .append('path')
+            .attr('d', RenderConstants.arrowheadShape)
+            .style('stroke', 'none')
+            .style('opacity', RenderConstants.LinkOpacity)
+    }
+
+    /**
+     * Appends a `g` element to a given `svg`, to be used as a container for other elements.
+     * @param svg A d3 `Selection` of an `svg` element to which the `g` element is added
+     * @param id The id of the `g` container appened to the `svg`
+     * @returns A d3 `Selection` of the `g` element appended to the given `svg` element
+     */
+    public appendGContainerElement(
+        svg: d3.Selection<SVGSVGElement, unknown, null | HTMLElement, unknown>,
```

Comment:
Can we alias `d3.Selection<SVGSVGElement, unknown, null | HTMLElement, unknown>` to its own type? I see it repeated a lot in this PR and it adds significant visual noise. Applies for any oft-used complex types such as this.

Reply:
Great idea.

Reply:
I've made an alias for the `d3.Selection<SVGSVGElement, unknown, null | HTMLElement, unknown>`', but we won't be able to alias the other `d3.Selection<...>`s because they're all slightly different. It is better without the SVG one though, as it was most common.